solvers/glmnet.py

```diff
@@ -38,18 +43,35 @@ def set_objective(self, X, y, lmbd, fit_intercept):
         self.X, self.y, self.lmbd = X, y, lmbd
         self.fit_intercept = fit_intercept
 
-        self.lmbd_max = np.max(np.abs(X.T @ y))
         self.glmnet = robjects.r['glmnet']
 
-    def run(self, n_iter):
-        fit_dict = {"lambda.min.ratio": self.lmbd / self.lmbd_max}
+    def run(self, tol):
+        # Even if maxit=0, glmnet can return non zero coefficients. To get the
+        # initial point on the curve, we set tol=0: this way, glmnet
+        # convergence check fails, and rather than returning the current
+        # iterate, it returns a 0 model.
+        if tol == INFINITY:
+            maxit = 0
+            thresh = 0
+        else:
+            maxit = 1_000_000
+            # we need thresh to decay fast, otherwise the objective curve can
+            # plateau before convergence
+            thresh = tol ** 1.8
+
+        # We fit with a single lambda because when computing a path, glmnet
+        # early stops the path based on an uncontrollable criterion. Fitting
+        # with a single lambda may be suboptimal if it is small, but there is
+        # no other way to force glmnet to solve for a prescribed lambda.
+        fit_dict = {"lambda": self.lmbd / len(self.y)}
```

Comment:
This really calls for a lasso_path benchmark :)

Reply:
well, even that: there is no way to make glmnet compute the full path